### PR TITLE
feat(smart_mpc_trajectory_follower): add smart_mpc_trajectory_follower

### PR DESCRIPTION
**Pull request by masayukiaino**
## Description

This PR contains the initial implementation of the smart_mpc_trajectory_follower package, a control algorithm that combines model predictive control and machine learning.
This package replaces trajectory_follower_node package and provides adaptive vehicle control.

## Tests performed

Tested on a real vehicle, autoware simulator and the python simulator included in the package.

## Interface changes

Not applicable.

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: launch/tier4_control_launch/launch/control.launch.py [hunk 1]
```diff
@@ -429,17 +468,24 @@ def add_launch_arg(name: str, default_value=None, description=None):
 
     # component
     add_launch_arg("use_intra_process", "false", "use ROS 2 component container communication")
-    add_launch_arg("use_multithread", "true", "use multithread")
+    add_launch_arg("use_multithread", "false", "use multithread")
```
Comment (kosuke55) [inline launch/tier4_control_launch/launch/control.launch.py:471]: could you please keep default behavior?

```suggestion
    add_launch_arg("use_multithread", "true", "use multithread")
```
?

File: launch/tier4_control_launch/launch/control.launch.py [hunk 2]
```diff
@@ -429,17 +468,24 @@ def add_launch_arg(name: str, default_value=None, description=None):
...
-        "component_container_mt",
-        condition=IfCondition(LaunchConfiguration("use_multithread")),
-    )
+
+    if True:
+        set_container_mt_executable = SetLaunchConfiguration(
+            "container_executable",
+            "component_container_mt",
+            condition=IfCondition(LaunchConfiguration("use_multithread")),
+        )
```
Comment (kosuke55) [inline launch/tier4_control_launch/launch/control.launch.py:488]: if condition is not needed?

```suggestion
    set_container_mt_executable = SetLaunchConfiguration(
        "container_executable",
        "component_container_mt",
        condition=IfCondition(LaunchConfiguration("use_multithread")),
    )
```
?

File: launch/tier4_control_launch/launch/control.launch.py [hunk 3]
```diff
@@ -345,21 +345,7 @@ def launch_setup(context, *args, **kwargs):
                 glog_component,
             ],
         )
-        control_validator_component = ComposableNode(
```
Comment (masayukiaino) [inline launch/tier4_control_launch/launch/control.launch.py]: Removed unnecessary case divisions.

File: launch/tier4_control_launch/launch/control.launch.py (hunk 1 above)
Comment (masayukiaino) [reply launch/tier4_control_launch/launch/control.launch.py]: Revisions were made as suggested.
https://github.com/autowarefoundation/autoware.universe/pull/6805/commits/a28212af7ecef811d9f324357f9a62d855792bea#diff-990cc6e7d5246c36ec9621c61ab42a868240d5a02154f9529258efb3af36e620L377-R378

File: launch/tier4_control_launch/launch/control.launch.py [hunk 4]
```diff
@@ -328,22 +330,37 @@ def launch_setup(context, *args, **kwargs):
...
+            package="rclcpp_components",
+            executable=LaunchConfiguration("container_executable"),
+            composable_node_descriptions=[
+                lane_departure_component,
+                shift_decider_component,
+                vehicle_cmd_gate_component,
+                operation_mode_transition_manager_component,
+                glog_component,
+            ],
+        )
```
Comment (kosuke55) [inline launch/tier4_control_launch/launch/control.launch.py:362]: please add exception

File: launch/tier4_control_launch/launch/control.launch.py [hunk 5]
```diff
@@ -430,11 +456,17 @@ def add_launch_arg(name: str, default_value=None, description=None):
...
+        "trajectory_follower_mode",
+        "trajectory_follower_node",
+        "Options for which trajectory_follower to use. Options: `trajectory_follower_node`, `smart_mpc_trajectory_follower`",
+    )
     set_container_executable = SetLaunchConfiguration(
         "container_executable",
         "component_container",
         condition=UnlessCondition(LaunchConfiguration("use_multithread")),
     )
+
```
Comment (kosuke55) [inline launch/tier4_control_launch/launch/control.launch.py:469]:
```suggestion
```

Comment (kosuke55) [reply launch/tier4_control_launch/launch/control.launch.py:473]: sorry it was added by pre-commits. it's ok

File: control/smart_mpc_trajectory_follower/README.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+In "mppi_ilqr" mode, the initial value of iLQR is given by the MPPI solution.
+
+> [!NOTE]
+> With the default settings, the "mppi" mode does not work well due to insufficient number of samples. Due to the limitation of using only the CPU, it is difficult to increase the number of samples sufficiently.
+
+To perform a simulation, run the following command:
+
+```bash
+ros2 launch autoware_launch planning_simulator.launch.xml map_path:=$HOME/autoware_map/sample-map-planning vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit trajectory_follower_mode:=smart_mpc_trajectory_follower
+```
```
Comment (TakaHoribe) [inline control/smart_mpc_trajectory_follower/README.md:47]:
```suggestion
To perform the smart_mpc controller in the planning simulation, run the following command:

```bash
ros2 launch autoware_launch planning_simulator.launch.xml map_path:=$HOME/autoware_map/sample-map-planning vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit trajectory_follower_mode:=smart_mpc_trajectory_follower
```

The control mode can be selected from "ilqr", "mppi", or "mppi_ilqr", and can be set as `mpc_parameter:system:mode` in [mpc_param.yaml](./smart_mpc_trajectory_follower/param/mpc_param.yaml).
In "mppi_ilqr" mode, the initial value of iLQR is given by the MPPI solution.

> [!NOTE]
> With the default settings, the performance of "mppi" mode is limited due to an insufficient number of samples. This issue is being addressed with ongoing work to introduce GPU support.
```

Comment (TakaHoribe) [reply control/smart_mpc_trajectory_follower/README.md:47]: .... suggestion seems not working well. Replase this part with 

---
To perform the smart_mpc controller in the planning simulation, run the following command:

```bash
ros2 launch autoware_launch planning_simulator.launch.xml map_path:=$HOME/autoware_map/sample-map-planning vehicle_model:=sample_vehicle sensor_model:=sample_sensor_kit trajectory_follower_mode:=smart_mpc_trajectory_follower
```

The control mode can be selected from "ilqr", "mppi", or "mppi_ilqr", and can be set as `mpc_parameter:system:mode` in [mpc_param.yaml](./smart_mpc_trajectory_follower/param/mpc_param.yaml).
In "mppi_ilqr" mode, the initial value of iLQR is given by the MPPI solution.

> [!NOTE]
> With the default settings, the performance of "mppi" mode is limited due to an insufficient number of samples. This issue is being addressed with ongoing work to introduce GPU support.
---

File: control/smart_mpc_trajectory_follower/README.md [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+```python
+lateral_error_visualize(dir_name=rosbag_nominal,ylim=[-1.2,1.2])
+lateral_error_visualize(dir_name=rosbag_trained,ylim=[-1.2,1.2])
+```
+
+The following results were obtained.
+
+<p><img src="images/lateral_error_nominal_model.png" width=712pix></p>
+<p><img src="images/lateral_error_trained_model.png" width=712pix></p>
```
Comment (TakaHoribe) [inline control/smart_mpc_trajectory_follower/README.md:145]: Just for preference. If you prefer

```suggestion
<div style="display: flex; justify-content: center; align-items: center;">
    <img src="images/lateral_error_nominal_model.png">
    <img src="images/lateral_error_trained_model.png">
</div>
```

File: control/smart_mpc_trajectory_follower/README.md [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+model_trainer.add_data_from_csv(save_dir)
+model_trainer.save_train_data(save_dir)
+model_trainer.get_trained_model(use_polynomial_reg=True)
+model_trainer.save_models(save_dir=save_dir)
+```
+
+This way, files `model_for_test_drive.pth` and `polynomial_reg_info.npz` are saved in `save_dir`.
+The following results were obtained.
+
+<p><img src="images/python_sim_lateral_error_nominal_model.png" width=712pix></p>
```
Comment (TakaHoribe) [inline control/smart_mpc_trajectory_follower/README.md:184]:
```suggestion
<p style="text-align: center;">
    <img src="images/python_sim_lateral_error_nominal_model.png" width="712px">
</p>
```

File: control/smart_mpc_trajectory_follower/README.md [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+
+```python
+load_dir = save_dir
+save_dir = "test_python_trained_sim"
+python_simulator.slalom_drive(save_dir=save_dir,load_dir=load_dir,use_trained_model=True,initial_error=initial_error)
+```
+
+The following results were obtained.
+
+<p><img src="images/python_sim_lateral_error_trained_model.png" width=712pix></p>
```
Comment (TakaHoribe) [inline control/smart_mpc_trajectory_follower/README.md:198]:
```suggestion
<p style="text-align: center;">
    <img src="images/python_sim_lateral_error_trained_model.png" width="712px">
</p>
```

**Review by TakaHoribe — COMMENTED** [review]
@masayukiaino some suggestions on readme

File: control/smart_mpc_trajectory_follower/smart_mpc_trajectory_follower/param/mpc_param.yaml [hunk 10]
```diff
@@ -0,0 +1,56 @@
+mpc_parameter:
+  system:
+    mode: ilqr
```
Comment (TakaHoribe) [inline control/smart_mpc_trajectory_follower/smart_mpc_trajectory_follower/param/mpc_param.yaml:3]:
```suggestion
    mode: ilqr  # option: ilqr, mppi, ilqr_mppi
```

File: control/smart_mpc_trajectory_follower/README.md (hunk 6 above)
Comment (masayukiaino) [reply control/smart_mpc_trajectory_follower/README.md]: Revised Note as suggested.
https://github.com/autowarefoundation/autoware.universe/pull/6805/commits/980abddb3e097ca16fddcd56a740693d9e8c88d4#diff-1b10418540e411525cff15236aca92b0e6b3f4f03ddfc58d723fb6980708f700R43

File: control/smart_mpc_trajectory_follower/README.md (hunk 8 above)
Comment (masayukiaino) [reply control/smart_mpc_trajectory_follower/README.md]: Revised as suggested.

https://github.com/autowarefoundation/autoware.universe/pull/6805/commits/980abddb3e097ca16fddcd56a740693d9e8c88d4#diff-1b10418540e411525cff15236aca92b0e6b3f4f03ddfc58d723fb6980708f700R188-R190

File: control/smart_mpc_trajectory_follower/smart_mpc_trajectory_follower/param/mpc_param.yaml (hunk 10 above)
Comment (masayukiaino) [reply control/smart_mpc_trajectory_follower/smart_mpc_trajectory_follower/param/mpc_param.yaml]: Revised as suggested.

https://github.com/autowarefoundation/autoware.universe/pull/6805/commits/980abddb3e097ca16fddcd56a740693d9e8c88d4#diff-615227b5a29e8f0cf76ed19c3b088d820c95fd78ac494f03ab5d3de07129e7d1R3

**Review by TakaHoribe — APPROVED** [review]
LGTM

File: launch/tier4_control_launch/launch/control.launch.py [hunk 11]
```diff
@@ -359,7 +381,7 @@ def launch_setup(context, *args, **kwargs):
         target_container="/control/control_container",
     )
 
-    # control validator checker
+    # control validator check
```
Comment (kosuke55) [inline launch/tier4_control_launch/launch/control.launch.py:384]:
```suggestion
    # control validator checker
```

**Pull request merged**